File: listeners/actions/handle_attention_request.py
from slack_bolt import Ack
from slack_sdk import WebClient
from logging import Logger
from .block_helpers import assemble_pr_message_blocks
from database import get_pr_by_id, add_pr_to_store

def handle_attention_request(ack: Ack, body: dict, client: WebClient, logger: Logger):
    try:
        ack()

        user_id = body["user"]["id"]
        action_id = body["actions"][0]["action_id"]
        pr_id = action_id.replace("attention_request_", "")
        logger.info(f"User {user_id} is attempting to request attention for PR {pr_id}")

        pr = get_pr_by_id(pr_id)

        if pr:
            if user_id in pr["attention_requests"]:
                del pr["attention_requests"][user_id]
                logger.info(f"User {user_id} removed their attention request from PR {pr_id}")
            else:
                pr["attention_requests"][user_id] = True
                logger.info(f"User {user_id} requested attention for PR {pr_id}")

                submitter_id = pr.get("submitter_id")
                if not submitter_id:
                    logger.error(f"Submitter ID for PR {pr_id} not found.")
                    return
                
                user_id_name = client.users_info(user=user_id)["user"]["real_name"]
                original_post_ts = pr.get('message_ts')

                # Send a DM to the submitter
                try:
                    client.chat_postMessage(
                        channel=submitter_id,
                        text=f"Attention requested by {user_id_name} for your PR {pr['name']}.",
                        blocks= [
                            {
                                "type": "section",
                                "text": {
                                    "type": "mrkdwn",
                                    "text": f"*Attention requested by {user_id_name} for your PR* *<{pr['link']}|{pr['name']}>*.\n<https://your-workspace.slack.com/archives/{pr['channel_id']}/p{original_post_ts.replace('.', '')}|View original post>"
                                }
                            }
                        ],
                        unfurl_links = False
                    )
                    logger.info(f"Sent attention request notification to submitter {submitter_id} for PR {pr_id}")
                except Exception as dm_error:
                    logger.error(f"Failed to send DM to submitter {submitter_id} for PR {pr_id}: {dm_error}")

            # Update the attention_requests in the database
            add_pr_to_store(pr)
            logger.info(f"PR {pr_id} attention requests updated in the store")

            # Update the original message with the new blocks
            blocks = assemble_pr_message_blocks(client, pr, user_id, logger)
            client.chat_update(
                channel=pr["channel_id"],
                ts=pr["message_ts"],
                blocks=blocks,
                text=blocks[0]["text"]["text"]
            )
            logger.info(f"Updated PR message in channel {pr['channel_id']} for PR {pr_id}")

        else:
            logger.error(f"PR {pr_id} not found in the store.")

    except Exception as e:
        logger.error(f"Error handling attention request for PR {pr_id}: {e}")

File: src/listeners/actions/HandleAttentionRequest.java
import com.slack.api.bolt.context.builtin.ActionContext;
import com.slack.api.bolt.handler.builtin.ActionHandler;
import com.slack.api.methods.SlackApiException;
import com.slack.api.model.block.LayoutBlock;
import org.slf4j.Logger;
import prreminderbot.BlockHelpers;
import prreminderbot.Database;

import java.io.IOException;
import java.util.List;
import java.util.Map;

public class HandleAttentionRequest {

    public static void handle(ActionRequest req, ActionContext ctx, Slack client, Logger logger) {
        try {
            ctx.ack();  // Acknowledge the action

            String userId = req.getPayload().getUser().getId();
            String actionId = req.getPayload().getActions().get(0).getActionId();
            String prId = actionId.replace("attention_request_", "");
            logger.info("User {} is attempting to request attention for PR {}", userId, prId);

            Map<String, Object> pr = Database.getPrById(prId);

            if (pr != null) {
                List<String> attentionRequests = (List<String>) pr.get("attention_requests");

                if (attentionRequests.contains(userId)) {
                    attentionRequests.remove(userId);
                    logger.info("User {} removed their attention request from PR {}", userId, prId);
                } else {
                    attentionRequests.add(userId);
                    logger.info("User {} requested attention for PR {}", userId, prId);

                    String submitterId = (String) pr.get("submitter_id");
                    if (submitterId == null) {
                        logger.error("Submitter ID for PR {} not found.", prId);
                        return;
                    }

                    String userName = client.methods().usersInfo(r -> r.user(userId)).getUser().getRealName();
                    String originalPostTs = (String) pr.get("message_ts");

                    // Send a DM to the submitter
                    try {
                        client.methods().chatPostMessage(r -> r
                                .channel(submitterId)
                                .text("Attention requested by " + userName + " for your PR " + pr.get("name") + ".")
                                .blocks(List.of(
                                        LayoutBlock.section(section -> section
                                                .text(text -> text.type("mrkdwn").text(
                                                        "*Attention requested by " + userName + " for your PR* *<" +
                                                                pr.get("link") + "|" + pr.get("name") + ">*.\n" +
                                                                "<https://your-workspace.slack.com/archives/" +
                                                                pr.get("channel_id") + "/p" +
                                                                originalPostTs.replace(".", "") +
                                                                "|View original post>"
                                                ))
                                        )
                                ))
                                .unfurlLinks(false)
                        );
                        logger.info("Sent attention request notification to submitter {} for PR {}", submitterId, prId);
                    } catch (IOException | SlackApiException e) {
                        logger.error("Failed to send DM to submitter {} for PR {}: {}", submitterId, prId, e.getMessage());
                    }
                }

                // Update the attention_requests in the database
                Database.addPrToStore(pr);
                logger.info("PR {} attention requests updated in the store", prId);

                // Update the original message with the new blocks
                List<LayoutBlock> blocks = BlockHelpers.assemblePrMessageBlocks(client, pr, userId, logger);
                client.methods().chatUpdate(r -> r
                        .channel((String) pr.get("channel_id"))
                        .ts((String) pr.get("message_ts"))
                        .blocks(blocks)
                        .text(blocks.get(0).getText().getText())
                );
                logger.info("Updated PR message in channel {} for PR {}", pr.get("channel_id"), prId);

            } else {
                logger.error("PR {} not found in the store.", prId);
            }

        } catch (Exception e) {
            logger.error("Error handling attention request for PR {}: {}", req.getPayload().getActions().get(0).getActionId().replace("attention_request_", ""), e.getMessage());
        }
    }
}
